### PR TITLE
feat(server): TodoWrite tool for claude-byok provider (#4051)

### PR DESCRIPTION
**Pull request by blamechris**
## Summary

Adds **TodoWrite** to the BYOK built-in toolset (#4051, deferred from epic #4047).

- Tool definition matches Claude Code's TodoWrite schema (`todos: Array<{ id, content, status, activeForm? }>`)
- Status is enum-validated: `pending | in_progress | completed`
- **Merge by id** — partial updates don't drop unrelated items (per #4051 acceptance)
- Store lives on `byok-session._todos` (Map keyed by id); resets on session destroy via GC
- **Atomic merge** — validation runs over the entire input array before any mutation, so a bad item mid-call doesn't half-apply
- Tool result echoes the full current list so the model sees unrelated items even after a partial call

## Test plan

- [x] 10 new TodoWrite subtests pass — merge semantics, atomicity, status validation, edge cases
- [x] Schema lock updated (`['Bash', 'Edit', 'Glob', 'Grep', 'Read', 'TodoWrite', 'Write']`)
- [x] Schema test asserts `todos` array + per-item `id/content/status` required
- [x] Full BYOK suite passes (104 tests across 30 suites)
- [x] byok-session tests still pass (30/30) after wiring `todoStore` through

Closes #4051.

### CODE REVIEW

File: packages/server/src/byok-tool-executor.js [hunk 1]
```diff
@@ -333,3 +338,68 @@ function shellQuote(s) {
...
+  const all = [...todoStore.values()]
+  const counts = { pending: 0, in_progress: 0, completed: 0 }
+  for (const t of all) counts[t.status]++
+
+  const lines = [`Todo list (${all.length} items): ${counts.in_progress} in progress, ${counts.pending} pending, ${counts.completed} completed`]
+  for (const t of all) {
+    const marker = t.status === 'completed' ? '[x]' : t.status === 'in_progress' ? '[~]' : '[ ]'
+    lines.push(`  ${marker} ${t.content} (${t.id})`)
+  }
+  return { content: lines.join('\n'), isError: false }
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/byok-tool-executor.js:604]: **FIX**

Fixed in `ce760f303`

**Change:** Added two caps — `TODOWRITE_MAX_ITEMS_RENDERED = 100` and `TODOWRITE_MAX_CONTENT_RENDERED = 200`. Long content strings are sliced + suffixed with `…`; overflow items are replaced with `… (showing first 100 of N; full list retained server-side)`. The full Map stays intact on the session.

```diff
-  const lines = [`Todo list (${all.length} items): ...`]
-  for (const t of all) {
-    lines.push(`  ${marker} ${t.content} (${t.id})`)
-  }
+  const visible = all.slice(0, TODOWRITE_MAX_ITEMS_RENDERED)
+  const lines = [header]
+  for (const t of visible) {
+    const content = t.content.length > TODOWRITE_MAX_CONTENT_RENDERED
+      ? t.content.slice(0, TODOWRITE_MAX_CONTENT_RENDERED) + '…' : t.content
+    lines.push(`  ${marker} ${content} (${t.id})`)
+  }
+  if (all.length > visible.length) {
+    lines.push(`  … (showing first ${visible.length} of ${all.length}; full list retained server-side)`)
+  }
```

Pinned with two new tests: 150-item list shows the 100-cap marker; 500-char content string is sliced.

File: packages/server/src/byok-tools.js [hunk 2]
```diff
@@ -11,11 +11,16 @@
...
- *   - TodoWrite — #4051
  *   - Task (subagent) — #4049
  *   - MCP — #4048
  */
 
+/**
+ * Valid TodoWrite status values. The model must emit one of these per
+ * item; anything else returns a tool error so the model can self-correct.
+ */
+export const TODO_STATUSES = new Set(['pending', 'in_progress', 'completed'])
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/byok-tools.js]: **FIX**

Fixed in `ce760f303`

**Change:** `TODO_STATUS_LIST` is now the single frozen array; the schema enum and the runtime Set both derive from it.

```diff
-export const TODO_STATUSES = new Set(['pending', 'in_progress', 'completed'])
+export const TODO_STATUS_LIST = Object.freeze(['pending', 'in_progress', 'completed'])
+export const TODO_STATUSES = new Set(TODO_STATUS_LIST)

-status: { type: 'string', enum: ['pending', 'in_progress', 'completed'], ... }
+status: { type: 'string', enum: TODO_STATUS_LIST, ... }
```

Pinned with `status enum and TODO_STATUSES Set share one source of truth` — fails if the two ever drift.

File: packages/server/src/byok-tool-executor.js [hunk 3]
```diff
@@ -93,9 +96,11 @@ export async function executeBuiltinTool({
...
       case 'Grep':
         return await runGrep({ input, cwd, cwdRealCache, cwdCacheTtl, signal })
+      case 'TodoWrite':
+        return runTodoWrite({ input, todoStore })
       default:
         return {
-          content: `Unknown tool: ${toolName}. The claude-byok provider ships with: Read, Write, Edit, Bash, Glob, Grep. MCP and other tools land in follow-up issues.`,
+          content: `Unknown tool: ${toolName}. The claude-byok provider ships with: Read, Write, Edit, Bash, Glob, Grep, TodoWrite. MCP and other tools land in follow-up issues.`,
           isError: true,
         }
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/src/byok-tool-executor.js]: **FIX**

Fixed in `ce760f303`

**Change:** Unknown-tool message now derives the shipped-tool list from `BUILTIN_TOOL_NAMES`. Adding a tool to byok-tools.js automatically reflects in the error.

```diff
-      default:
-        return {
-          content: `Unknown tool: ${toolName}. The claude-byok provider ships with: Read, Write, Edit, Bash, Glob, Grep, TodoWrite. MCP and other tools land in follow-up issues.`,
-          isError: true,
-        }
+      default: {
+        const known = [...BUILTIN_TOOL_NAMES].sort().join(', ')
+        return {
+          content: `Unknown tool: ${toolName}. The claude-byok provider ships with: ${known}. MCP and other tools land in follow-up issues.`,
+          isError: true,
+        }
+      }
```

Pinned with `error message lists every BUILTIN_TOOL name (drift guard)`.
